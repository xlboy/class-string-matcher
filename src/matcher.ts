import {
  htmlLexer,
  jsxLexer,
  solidLexer,
  vueDynamicLexer,
  vuePureLexer,
} from './lexer';
import moo from 'moo';

const SUPPORTED_LANGUAGE_IDS = ['jsx', 'html', 'vue', 'solid'] as const;
export type LanguageId = (typeof SUPPORTED_LANGUAGE_IDS)[number];

export interface ClassNode {
  text: string;
  pos: { s: number; e: number };
}

interface VueOptions {
  /**
   * Indicates whether the text is within a dynamic context where expressions are allowed.
   * This includes scenarios like:
   * - Vue `:class="..."`
   * - Or similar constructs in other frameworks where **dynamic values** can be used for class names.
   */
  inDynamicContext?: boolean;
}

export function classStringMatcher(
  text: string,
  languageId: 'vue',
  options: VueOptions,
): ClassNode[];
export function classStringMatcher(
  text: string,
  languageId: Exclude<LanguageId, 'vue'>,
): ClassNode[];
export function classStringMatcher(
  text: string,
  languageId: LanguageId,
  options?: VueOptions,
): ClassNode[] {
  if (!SUPPORTED_LANGUAGE_IDS.includes(languageId))
    throw new Error(`Unknown languageId: ${languageId}`);

  let lexer: moo.Lexer;
  switch (languageId) {
    case 'jsx':
      lexer = jsxLexer;
      break;
    case 'vue':
      lexer = options?.inDynamicContext ? vueDynamicLexer : vuePureLexer;
      break;
    case 'solid':
      lexer = solidLexer;
      break;
    case 'html':
    default:
      lexer = htmlLexer;
      break;
  }
  lexer.reset(text);

  const parsed = parser(lexer);
  const classNodes: ClassNode[] = parsed.map(v => {
    if (v.isEmpty) {
      return {
        text: '',
        pos: { s: v.tokens[1].offset, e: v.tokens[1].offset },
      };
    }
    return {
      text: v.tokens.map(t => t.text).join(''),
      pos: { s: v.tokens[0].offset, e: v.tokens.at(-1)!.offset },
    };
  });

  return classNodes;
}

function parser(lexer: moo.Lexer) {
  const classlist: Array<{ tokens: moo.Token[]; isEmpty: boolean }> = [];
  let currentParts: moo.Token[] = [];

  // 加 try 是为了让后面无效 token 不会报错（但前面的有效 token 会被保留）
  try {
    let t: moo.Token;
    let lastT: moo.Token | undefined;
    while ((t = lexer.next()!)) {
      switch (t.type) {
        case 'ignore':
        case 'start':
          break;

        case 'children':
        case 'end': {
          if (currentParts.length > 0) {
            classlist.push({ tokens: currentParts, isEmpty: false });
          } else if (lastT?.text === t.text) {
            classlist.push({ tokens: [lastT, t], isEmpty: true });
          }
          currentParts = [];
          break;
        }

        case 'content': {
          currentParts.push(t);
          break;
        }

        default:
          throw new Error(`Unexpected token: ${t.type}`);
      }
      lastT = t;
    }
  } catch (_) {}

  return classlist;
}
