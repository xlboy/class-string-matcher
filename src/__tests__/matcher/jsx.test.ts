import { classStringMatcher } from '../../matcher';
import { expect, it } from 'vitest';

const codes = [
  `"">`,
  `"text-red text-lg"`,
  `'text-blue  ...'`,
  `'a b c d ', ../...`,
  `{a: 'b c d'}`,
  `{"a b c d"}`,
  "`text-red ${true && 'text-lg'}`",
  "`text-red ${true && 'text-lg'} ${false && 'text-sm'}`",
  "{'bbb' + \"ccc\" + `ddd` + `eeee ${true && 'fff'} 那天车窗起雾`} >你好</div>",
  `(a + b + c)`,
  `(a + "text-red" + c)`,
  `(a + {b: 'text-red'} + c + 'text-lg', 'text-sm', \`haha- \${false || 'hacker'}\`)`,
  `(a + {b: 'text-red'} + ["test"])`,
];

it.each(codes)('%s', code => {
  expect(classStringMatcher(code, 'jsx')).toMatchSnapshot();
});
