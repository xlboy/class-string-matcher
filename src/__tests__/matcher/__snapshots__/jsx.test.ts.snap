// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`""> 1`] = `
[
  {
    "pos": {
      "e": 1,
      "s": 1,
    },
    "text": "",
  },
]
`;

exports[`"text-red text-lg" 1`] = `
[
  {
    "pos": {
      "e": 16,
      "s": 1,
    },
    "text": "text-red text-lg",
  },
]
`;

exports[`'a b c d ', ../... 1`] = `
[
  {
    "pos": {
      "e": 8,
      "s": 1,
    },
    "text": "a b c d ",
  },
]
`;

exports[`'text-blue  ...' 1`] = `
[
  {
    "pos": {
      "e": 14,
      "s": 1,
    },
    "text": "text-blue  ...",
  },
]
`;

exports[`(a + "text-red" + c) 1`] = `
[
  {
    "pos": {
      "e": 13,
      "s": 6,
    },
    "text": "text-red",
  },
]
`;

exports[`(a + {b: 'text-red'} + ["test"]) 1`] = `
[
  {
    "pos": {
      "e": 17,
      "s": 10,
    },
    "text": "text-red",
  },
  {
    "pos": {
      "e": 28,
      "s": 25,
    },
    "text": "test",
  },
]
`;

exports[`(a + {b: 'text-red'} + c + 'text-lg', 'text-sm', \`haha- \${false || 'hacker'}\`) 1`] = `
[
  {
    "pos": {
      "e": 17,
      "s": 10,
    },
    "text": "text-red",
  },
  {
    "pos": {
      "e": 34,
      "s": 28,
    },
    "text": "text-lg",
  },
  {
    "pos": {
      "e": 45,
      "s": 39,
    },
    "text": "text-sm",
  },
  {
    "pos": {
      "e": 55,
      "s": 50,
    },
    "text": "haha- ",
  },
  {
    "pos": {
      "e": 73,
      "s": 68,
    },
    "text": "hacker",
  },
]
`;

exports[`(a + b + c) 1`] = `[]`;

exports[`\`text-red \${true && 'text-lg'} \${false && 'text-sm'}\` 1`] = `
[
  {
    "pos": {
      "e": 9,
      "s": 1,
    },
    "text": "text-red ",
  },
  {
    "pos": {
      "e": 27,
      "s": 21,
    },
    "text": "text-lg",
  },
  {
    "pos": {
      "e": 30,
      "s": 30,
    },
    "text": " ",
  },
  {
    "pos": {
      "e": 49,
      "s": 43,
    },
    "text": "text-sm",
  },
]
`;

exports[`\`text-red \${true && 'text-lg'}\` 1`] = `
[
  {
    "pos": {
      "e": 9,
      "s": 1,
    },
    "text": "text-red ",
  },
  {
    "pos": {
      "e": 27,
      "s": 21,
    },
    "text": "text-lg",
  },
]
`;

exports[`{"a b c d"} 1`] = `
[
  {
    "pos": {
      "e": 8,
      "s": 2,
    },
    "text": "a b c d",
  },
]
`;

exports[`{'bbb' + "ccc" + \`ddd\` + \`eeee \${true && 'fff'} 那天车窗起雾\`} >你好</div> 1`] = `
[
  {
    "pos": {
      "e": 4,
      "s": 2,
    },
    "text": "bbb",
  },
  {
    "pos": {
      "e": 12,
      "s": 10,
    },
    "text": "ccc",
  },
  {
    "pos": {
      "e": 20,
      "s": 18,
    },
    "text": "ddd",
  },
  {
    "pos": {
      "e": 30,
      "s": 26,
    },
    "text": "eeee ",
  },
  {
    "pos": {
      "e": 44,
      "s": 42,
    },
    "text": "fff",
  },
  {
    "pos": {
      "e": 53,
      "s": 47,
    },
    "text": " 那天车窗起雾",
  },
]
`;

exports[`{a: 'b c d'} 1`] = `
[
  {
    "pos": {
      "e": 9,
      "s": 5,
    },
    "text": "b c d",
  },
]
`;
